src/pudl/load/parquet.py

```diff
@@ -71,11 +30,19 @@ def epacems_to_parquet(df, root_path):
         None
 
     """
+    epacems_pyarrow_schema = (
+        pudl.metadata.classes.Package.from_resource_ids(
+            pudl.metadata.RESOURCE_METADATA
+        )
+        .get_resource("hourly_emissions_epacems")
```

Comment:
If the package were pre-computed, I agree that `package.get_resource('hourly_emissions_epacems')` is the way to go. Otherwise, could you just do `Resource.from_id('hourly_emissions_epacems')`?

p.s. Why are the field encoders loaded by `Package.from_resource_ids()` (https://github.com/catalyst-cooperative/pudl/blob/9e6a23f0f47d593571177de8e86e4954681af032/src/pudl/metadata/classes.py#L1556) and not `Resource.from_id()`, when as far as I can tell, this depends only on the foreign keys assigned to that resource?

Reply:
`Resource.dict_from_id()` loads foreign keys from `FOREIGN_KEYS`, which is currently computed on load (perhaps another candidate for a cached build function). By the end of that method, `schema['foreign_keys']` is populated, so I think this code block should be moved there.

While you are at it, I might suggest rewriting:

```py
if len(fk["fields"]) == 1 and encoder:
    # fk["fields"] is a one element list, get the one element:
    field = fk["fields"][0]
    field_names = [f["name"] for f in resource["schema"]["fields"]]
    idx = field_names.index(field)
    resource["schema"]["fields"][idx]["encoder"] = encoder
```

as:

```py
if len(fk["fields"]) == 1 and encoder:
    # fk["fields"] is a one element list, get the one element:
    field = fk["fields"][0]
    for f in resource["schema"]["fields"]:
        if f["name"] == field:
            f["encoder"] = encoder
            break
```

Reply:
Ah yeah that's clearer.

Reply:
If I move the Encoder assignment into `Resource.from_id()` then all the column encoders get set to None. So something hasn't been assigned somewhere. Is this what you were thinking?

```python
    @classmethod
    def from_id(cls, x: str) -> "Resource":
        """Construct from PUDL identifier (`resource.name`)."""
        resource = cls(**cls.dict_from_id(x))
        # Foreign key relationships determine the set of codes to use
        for fk in resource["schema"]["foreign_keys"]:
            # Only referenced tables with an associated encoder indicate
            # that the column we're looking at should have an encoder
            # attached to it. All of these FK relationships must have simple
            # single-column keys.
            encoder = Encoder.dict_from_id(fk["reference"]["resource"])
            if len(fk["fields"]) == 1 and encoder:
                # fk["fields"] is a one element list, get the one element:
                field = fk["fields"][0]
                for f in resource["schema"]["fields"]:
                    if f["name"] == field:
                        f["encoder"] = encoder
                        break
        return resource
```

I don't think I'm understanding the roles of `dict_from_id()` and `from_id()` -- what things are supposed to happen when the object is represented by a big nested dictionary, and what things are supposed to be the instantiated class. Or maybe even why these two separate representations exist.

Reply:
In your code, `resource` is a class instance (`cls(**cls.dict_from_id(x))`), not a `dict`, so you would need to use dot-notation to access its properties (e.g. `resource.schema.foreign_keys`). Otherwise you'd get:

> TypeError: 'Resource' object is not subscriptable

But this should go into `Resource.dict_from_id()`. The reason is that `dict_from_id()` is where all of the PUDL-specific (meta)data is assembled, often by calling subclass `dict_from_id()`. It is useful to defer all class instantiation to the very end, since otherwise you raise an error on the first invalid subclass data, rather than receiving a complete nested error report. Plus I see value in preparing the data from which the classes are instantiated without the help of the classes themselves.

Reply:
Alright I think this works now.